### PR TITLE
Support for SARIF suppressed issues and remove rule details

### PR DESCRIPTION
**Pull request by Gby56**
Fixes both https://github.com/azu/security-alert/issues/12 and https://github.com/azu/security-alert/issues/11
This adds simple flags like `--ruleDetails` in the cli, hiding the rule JSON by default.
This also separates the output in separate sections, first the results, then the suppressed results (like adding special comments in code to ignore the finding)

I had build issues with types, I hope the package.json is not broken, my github action kept failing with 
```
 lerna info Executing command in 6 packages: "yarn run build"
lerna ERR! yarn run build exited 2 in '@security-alert/share'
lerna ERR! yarn run build stdout:
$ cross-env NODE_ENV=production tsc -p .
Error: src/VulnerabilityAlerts.ts(1,25): error TS2307: Cannot find module '@octokit/graphql' or its corresponding type declarations.
Error: src/fetchPackageDetails.ts(1,25): error TS2307: Cannot find module '@octokit/rest' or its corresponding type declarations.
Error: src/fetchPackageDetails.ts(7,18): error TS2580: Cannot find name 'require'. Do you need to install type definitions for node? Try `npm i --save-dev @types/node`.
Error: src/fetchPackageDetails.ts(49,39): error TS2580: Cannot find name 'Buffer'. Do you need to install type definitions for node? Try `npm i --save-dev @types/node`.
Error: src/parser/package-lock.json.ts(1,29): error TS2307: Cannot find module '@npm/types' or its corresponding type declarations.
Error: src/parser/package.json.ts(1,29): error TS2307: Cannot find module '@npm/types' or its corresponding type declarations.
```
I also fixed a few tsc issues
```
lerna ERR! yarn run build exited 2 in '@security-alert/sarif-to-issue'
lerna ERR! yarn run build stdout:
$ tsc -p . && tsc --project ./tsconfig.module.json
src/cli.ts(98,17): error TS2322: Type 'string | undefined' is not assignable to type 'string'.
  Type 'undefined' is not assignable to type 'string'.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

```
$ tsc -p . && tsc --project ./tsconfig.module.json
lerna ERR! yarn run build exited 2 in '@security-alert/list-alerts'
lerna ERR! yarn run build stdout:
$ cross-env NODE_ENV=production tsc -p .
src/cli.ts(43,27): error TS2532: Object is possibly 'undefined'.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

```
lerna ERR! yarn run build exited 2 in '@security-alert/share'
lerna ERR! yarn run build stdout:
$ cross-env NODE_ENV=production tsc -p .
src/fetchPackageDetails.ts(6,1): error TS6133: 'OctokitResponse' is declared but its value is never read.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
lerna ERR! yarn run build stderr:
error Command failed with exit code 2.
lerna ERR! yarn run build exited 2 in '@security-alert/share'
```

```
lerna ERR! yarn run build exited 2 in '@security-alert/share'
lerna ERR! yarn run build stdout:
$ cross-env NODE_ENV=production tsc -p .
src/fetchPackageDetails.ts(45,32): error TS2339: Property 'type' does not exist on type '{ type: string; size: number; name: string; path: string; content?: string | undefined; sha: string; url: string; git_url: string | null; html_url: string | null; download_url: string | null; _links: { ...; }; }[] | { ...; } | { ...; } | { ...; }'.
  Property 'type' does not exist on type '{ type: string; size: number; name: string; path: string; content?: string | undefined; sha: string; url: string; git_url: string | null; html_url: string | null; download_url: string | null; _links: { ...; }; }[]'.
src/fetchPackageDetails.ts(48,32): error TS2339: Property 'encoding' does not exist on type '{ type: string; size: number; name: string; path: string; content?: string | undefined; sha: string; url: string; git_url: string | null; html_url: string | null; download_url: string | null; _links: { ...; }; }[] | { ...; } | { ...; } | { ...; }'.
  Property 'encoding' does not exist on type '{ type: string; size: number; name: string; path: string; content?: string | undefined; sha: string; url: string; git_url: string | null; html_url: string | null; download_url: string | null; _links: { ...; }; }[]'.
src/fetchPackageDetails.ts(49,60): error TS2339: Property 'content' does not exist on type '{ type: string; size: number; name: string; path: string; content?: string | undefined; sha: string; url: string; git_url: string | null; html_url: string | null; download_url: string | null; _links: { ...; }; }[] | { ...; } | { ...; } | { ...; }'.
  Property 'content' does not exist on type '{ type: string; size: number; name: string; path: string; content?: string | undefined; sha: string; url: string; git_url: string | null; html_url: string | null; download_url: string | null; _links: { ...; }; }[]'.
src/fetchPackageDetails.ts(51,46): error TS2339: Property 'html_url' does not exist on type '{ type: string; size: number; name: string; path: string; content?: string | undefined; sha: string; url: string; git_url: string | null; html_url: string | null; download_url: string | null; _links: { ...; }; }[] | { ...; } | { ...; } | { ...; }'.
  Property 'html_url' does not exist on type '{ type: string; size: number; name: string; path: string; content?: string | undefined; sha: string; url: string; git_url: string | null; html_url: string | null; download_url: string | null; _links: { ...; }; }[]'.
src/fetchPackageDetails.ts(55,46): error TS2339: Property 'html_url' does not exist on type '{ type: string; size: number; name: string; path: string; content?: string | undefined; sha: string; url: string; git_url: string | null; html_url: string | null; download_url: string | null; _links: { ...; }; }[] | { ...; } | { ...; } | { ...; }'.
  Property 'html_url' does not exist on type '{ type: string; size: number; name: string; path: string; content?: string | undefined; sha: string; url: string; git_url: string | null; html_url: string | null; download_url: string | null; _links: { ...; }; }[]'.
src/fetchPackageDetails.ts(59,66): error TS2339: Property 'type' does not exist on type '{ type: string; size: number; name: string; path: string; content?: string | undefined; sha: string; url: string; git_url: string | null; html_url: string | null; download_url: string | null; _links: { ...; }; }[] | { ...; } | { ...; } | { ...; }'.
  Property 'type' does not exist on type '{ type: string; size: number; name: string; path: string; content?: string | undefined; sha: string; url: string; git_url: string | null; html_url: string | null; download_url: string | null; _links: { ...; }; }[]'.
src/fetchPackageDetails.ts(59,88): error TS2339: Property 'encoding' does not exist on type '{ type: string; size: number; name: string; path: string; content?: string | undefined; sha: string; url: string; git_url: string | null; html_url: string | null; download_url: string | null; _links: { ...; }; }[] | { ...; } | { ...; } | { ...; }'.
  Property 'encoding' does not exist on type '{ type: string; size: number; name: string; path: string; content?: string | undefined; sha: string; url: string; git_url: string | null; html_url: string | null; download_url: string | null; _links: { ...; }; }[]'.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
lerna ERR! yarn run build stderr:
error Command failed with exit code 2.
lerna ERR! yarn run build exited 2 in '@security-alert/share'
```

### CODE REVIEW

**Comment by Gby56** [comment]
@azu Maybe I could add another cli flag to include or exclude suppressed findings in the comment ? If the scan is covering a lot of things, it might find too maybe suppressed findings and exceed the Github API size limit... ?
Overall it would be good to make a multi-comment implementation, if 1 comment is too big, split in 2

File: packages/create-issue/package.json [hunk 1]
```diff
@@ -41,10 +41,9 @@
     "test": "# jest",
     "watch": "tsc -p . --watch"
   },
-
-
-
   "devDependencies": {
+    "@octokit/graphql": "^4.8.0",
+    "@octokit/rest": "^18.12.0",
```
Comment (azu) [inline packages/create-issue/package.json:46]: These is dulicated. We need to remove these from devDependencies.

**Comment by azu** [comment]
> Maybe I could add another cli flag to include or exclude suppressed findings in the comment ?

I agree to add the flag for suppressed results.
I'm ok that the default value of suppressed results flag is `true`.
It allows user to disable the flag like `--no-suppressedResult` ([meow](https://github.com/sindresorhus/meow) support `--no-<flag>` prefix).

> Overall it would be good to make a multi-comment implementation, if 1 comment is too big, split in 2 comments, 

I'm worried that multi-comment implementation make complex.

> or just upload an artifact ?

GitHub support SARIF uploads.
So, I not want to implement uploading feature.

[Uploading a SARIF file to GitHub - GitHub Docs](https://docs.github.com/en/code-security/code-scanning/integrating-with-code-scanning/uploading-a-sarif-file-to-github)

File: packages/create-issue/package.json (hunk 1 above)
Comment (Gby56) [reply packages/create-issue/package.json]: Fixed :)

**Comment by Gby56** [comment]
I agree with your cli, I'll look into this and push to my branch to update the PR 👍 
Also agree on the multi-comment implem, it's complex, and uploading the report in the comment kinda defeats the purpose (but to be fair, you have to pay to get the Gihub SARIF upload I think, it's the Advanced Security feature I think)

**Comment by Gby56** [comment]
Done ! the flag should have the correct default value to true so that suppressed issues are always included.

File: .github/workflows/test-pr.yml [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+      - name: yarn install and build
+        run: |
+          npm install --global cross-env & \
+          yarn install & \
+          yarn run build
+      - name: try the build
+        run: |
+          cat packages/sarif-to-markdown/test/snapshots/suppressed-finding/input.json & \
+          ./packages/sarif-to-comment/bin/cmd.js packages/sarif-to-markdown/test/snapshots/suppressed-finding/input.json \
+          --commentUrl https://github.com/Gby56/security-alert/issues/1 \
```
Comment (azu) [inline .github/workflows/test-pr.yml:32]: I want to move this test CI to this repository.
I've invited you to this repository.
Free feel to use this repository for testing.

**Review by azu — APPROVED** [review]
LGTM.

Thanks to contribute and welcome to collaborator!

**Pull request merged**